In [2]:
import numpy as np

In [3]:
#Correct Running Code for Simpsons 1/3 in Three Dimensions

def g(x, y, z):
    return np.exp(x*y) + np.sin(z)

def SimpsonsTriple13(g, ax, bx, ay, by, az, bz,  nx, ny, nz):
    hx = (bx - ax) / nx
    hy = (by - ay) / ny
    hz = (bz - az) / nz
    integration = 0.0

    for i in range(0,nx+1):
        for j in range(0, ny+1):
            for k in range (0, nz+1):
                x = ax + i * hx
                y = ay + j * hy
                z = az + k * hz

                if i==0 or i==nx:
                    coefficientx =  1.0
                elif i%2 == 0:
                    coefficientx =  2.0
                else:
                    coefficientx =  4.0

                if j==0 or j==ny:
                    coefficienty =  1.0
                elif j%2 ==0:
                    coefficienty =  2.0
                else:
                    coefficienty =  4.0

                if k==0 or k==nz:
                    coefficientz = 1.0
                elif k%2 == 0:
                    coefficientz = 2.0
                else:
                    coefficientz = 4.0

                coefficient = coefficientx * coefficienty * coefficientz
                integration += coefficient*g(x,y,z)


    integration *= (hx * hy * hz) / 27.0

    return integration

# Define the region of integration [ax, bx] x [ay, by] x [az,bz]
ax, bx = 0, 2      # Limits in x
ay, by = 0, 1      # Limits in y
az, bz = 0, 1      # Limits in z
nx, ny, nz  = 160,160,160   # Number of divisions x and y and z

result = SimpsonsTriple13(g, ax, bx, ay, by, az, bz,  nx, ny, nz)
print("The 3 dimensional approximation using Simpson's 1/3 Rule is: ", result)

The 3 dimensional approximation using Simpson's 1/3 Rule is:  4.603266899095987


Lower Limit of x is equal to 0
Upper Limit of x is equal to 2
Lower Limit of y is equal to 0
Upper Limit of y is equal to 1
Lower Limit of z is equal to 0
Upper Limit of z is equal to 1
- nx = ny = nz = 10 the value is 4.603285989354729
- nx = ny = nz = 20 the value is 4.6032680948587466
- nx = ny = nz = 40 the value is 4.603266973605006
- nx = ny = nz = 80 the value is 4.603266903480031
- nx = ny = nz = 160 the value is 4.603266899095987

The true value is 4.6032668988038

Therefore, the error vector would be
[.0000490905509, .0000011960549, .0000000748012, .0000000046762, .0000000002921]

In [4]:
ErrorVec = [.0000490905509, .0000011960549, .0000000748012, .0000000046762, .0000000002921]
def GetConvergence(vec,factor=2):
    for i in range(len(vec)-1):
        print("Convergence = ", (np.log(vec[i]/vec[i+1]))/np.log(factor))
GetConvergence(ErrorVec)

Convergence =  5.359089839716421
Convergence =  3.9990783870682964
Convergence =  3.999652874730167
Convergence =  4.000802371644955


In [9]:
#Correct Running Code for Simpsons 1/3 in Three Dimensions

def g(x, y, z):
    return np.exp(x*y) + np.sin(z)

def SimpsonsTriple38(g, ax, bx, ay, by, az, bz,  nx, ny, nz):
    hx = (bx - ax) / nx
    hy = (by - ay) / ny
    hz = (bz - az) / nz
    integration = 0.0

    for i in range(0,nx+1):
        for j in range(0, ny+1):
            for k in range (0, nz+1):
                x = ax + i * hx
                y = ay + j * hy
                z = az + k * hz

                if i==0 or i==nx:
                    coefficientx =  1.0
                elif i%3 == 0:
                    coefficientx =  2.0
                else:
                    coefficientx =  3.0

                if j==0 or j==ny:
                    coefficienty =  1.0
                elif j%3 ==0:
                    coefficienty =  2.0
                else:
                    coefficienty =  3.0

                if k==0 or k==nz:
                    coefficientz = 1.0
                elif k%3 == 0:
                    coefficientz = 2.0
                else:
                    coefficientz = 3.0

                coefficient = coefficientx * coefficienty * coefficientz
                integration += coefficient*g(x,y,z)


    integration *= (27.0*hx * hy * hz) / 512.0

    return integration

# Define the region of integration [ax, bx] x [ay, by] x [az,bz]
ax, bx = 0, 2      # Limits in x
ay, by = 0,1      # Limits in y
az, bz = 0,1      # Limits in z
nx, ny, nz  = 144,144,144  # Number of divisions x and y and z and they must be divisible by 3

result = SimpsonsTriple38(g, ax, bx, ay, by, az, bz,  nx, ny, nz)
print("The 3 dimensional approximation using Simpson's 3/8 Rule is: ", result)

The 3 dimensional approximation using Simpson's 3/8 Rule is:  4.603266899806264


Lower Limit of x is equal to 0
Upper Limit of x is equal to 2
Lower Limit of y is equal to 0
Upper Limit of y is equal to 1
Lower Limit of z is equal to 0
Upper Limit of z is equal to 1
- nx = ny = nz = 9 the value is 4.603332063614332
- nx = ny = nz = 18 the value is 4.6032709954345865
- nx = ny = nz = 36 the value is 4.6032671552423565
- nx = ny = nz = 72 the value is 4.603266914837851
- nx = ny = nz = 144 the value is 4.603266899806264

The true value is 4.6032668988038

Therefore, the error vector would be
[.0000651648105, .0000040966307, .0000002564385, .000000016034, .0000000010024]

In [10]:
ErrorVec2 = [.0000651648105, .0000040966307, .0000002564385, .000000016034, .0000000010024]
def GetConvergence(vec,factor=2):
    for i in range(len(vec)-1):
        print("Convergence = ", (np.log(vec[i]/vec[i+1]))/np.log(factor))
GetConvergence(ErrorVec2)

Convergence =  3.991583359489186
Convergence =  3.997753063554139
Convergence =  3.9994065905602643
Convergence =  3.999604154461797
